lib/src/features/feature_renderer.dart

```diff
@@ -8,9 +12,89 @@ import 'polygon_renderer.dart';
 import 'symbol_line_renderer.dart';
 import 'symbol_point_renderer.dart';
 
+class FeatureRendererContext implements Context {
+  FeatureRendererContext(this._context, this.pixelsPerTileUnit);
+
+  final Context _context;
+  final double pixelsPerTileUnit;
+
+  @override
+  LabelSpace get labelSpace => _context.labelSpace;
+
+  @override
+  set labelSpace(LabelSpace _labelSpace) => _context.labelSpace = _labelSpace;
+
+  @override
+  Canvas get canvas => _context.canvas;
+
+  @override
+  FeatureDispatcher get featureRenderer => _context.featureRenderer;
+
+  @override
+  Logger get logger => _context.logger;
+
+  @override
+  Tile? tile(String sourceId) => _context.tile(sourceId);
+
+  @override
+  Rect get tileClip => _context.tileClip;
+
+  @override
+  Rect get tileSpace => _context.tileSpace;
+
+  @override
+  Tileset get tileset => _context.tileset;
+
+  @override
+  double get zoom => _context.zoom;
+
+  @override
+  double get zoomScaleFactor => _context.zoomScaleFactor;
+
+  double widthFromPixelToTile(double value) {
+    return value / pixelsPerTileUnit;
+  }
+
+  Offset pointFromTileToPixels(Offset point) {
+    return point * pixelsPerTileUnit;
+  }
+
+  Size sizeFromTileToPixels(Size point) {
+    return point * pixelsPerTileUnit;
+  }
+
+  Rect rectFromTileToPixels(Rect rect) {
+    return pointFromTileToPixels(rect.topLeft) &
+        sizeFromTileToPixels(rect.size);
+  }
+
+  bool isPathWithinTileClip(Path path) {
+    return tileClip.overlaps(rectFromTileToPixels(path.getBounds()));
```

Comment:
This will get called quite a lot while rendering. How about having a `pixelTileClip` member that is converted from tile to pixels once, then we can check `overlaps` directly without converting.

Reply:
Done